packages/lit-element/demo/controllers.html

```diff
@@ -0,0 +1,433 @@
+        }
+        super.onConnected();
+      }
+
+      onListen(e) {
+        console.info(this.constructor.name, e.type);
+        this.element.message = e.type;
+        this.requestUpdate();
+      }
+
+    }
+
+    class MyHelper extends UpdatingController {
+
+      constructor(host) {
+        super(host);
+        this.listeners = [
+          new MyListener(this, 'click'),
+          new MyListener(this, 'mousedown'),
+          new MyListener(this, 'mouseup')
+        ];
+        this.logger = new MyLogger(this);
+      }
+
+    }
+
+    class MyCounter extends UpdatingController {
+
+      _count = 0;
+
+      constructor(host) {
+        super(host);
+        this.helper = new MyHelper(this);
+      }
+
+      get count() {
+        return this._count;
+      }
+
+      set count(value) {
+        this._count = value;
+        this.requestUpdate();
+      }
+
+    }
+
+    const addEvent = 'add-context';
+    const removeEvent = 'remove-context';
+
+    class Provider extends UpdatingController {
+
+        consumers = new Set();
+        key = null;
+        _value = null;
+
+        constructor(host, value) {
+          super(host);
+          this.value = value;
+          this.element.addEventListener(addEvent, (e) => this.onAddContext(e));
+          this.element.addEventListener(removeEvent, (e) => this.onRemoveContext(e));
+        }
+
+        onAddContext(e) {
+          const consumer = e.detail;
+          if (this.key !== consumer.key) {
+            return;
+          }
+          this.consumers.add(consumer);
+          consumer.value = this.value;
+          consumer.provider = this;
+          e.stopPropagation();
+        }
+
+        onRemoveContext(e) {
+          const consumer = e.detail;
+          if (this.key !== consumer.key) {
+            return;
+          }
+          this.consumers.delete(consumer);
+          consumer.provider = null;
+          e.stopPropagation();
+        }
+
+        get value() {
+          return this._value;
+        }
+
+        set value(value) {
+          this._value = value;
+          for (const consumer of this.consumers) {
+            consumer.value = this.value;
+          }
+        }
+
+      }
+
+      class Consumer extends UpdatingController {
+
+        _value = null;
+        provider = null;
+        key = null;
+
+        onConnected() {
+          this.element.dispatchEvent(new CustomEvent(addEvent, {detail: this, composed: true, bubbles: true}));
+        }
+
+        onDisconnected() {
+          if (this.provider) {
+            this.provider.element.dispatchEvent(new CustomEvent(removeEvent, {detail: this}));
+          }
+        }
+
+        get value() {
+          return this._value;
+        }
+
+        set value(value) {
+          this._value = value;
+          this.requestUpdate();
+        }
+
+      }
+
+    const createContext = (initialValue, key, ProviderBase = Provider, ConsumerBase = Consumer) => {
```

Comment:
There's an alternative API for context where we use directives, which I think we should at least explore.  Because context is tree based, directives seem more semantic and more closely match React, but doesn't lend itself as well to some patterns.

Roughly:

```html
<my-app ${provide(ThemeContext, 'dark')}>
  ...
  <my-button>
    #shadow-root
      <button class=${consume(ThemeContext, value => value)}>
```

The above strawman would require bringing back async commit ability to directives, and getting the initial value for consumption requires being connected to the tree. But API wise, this seems nicer to me.

Reply:
Context is intended to basically be similar to DI. I don't see how the desire "I want something provided to me from somewhere up the tree" makes sense as a directive?!?